.storybook/decorators/PropCombinations/README.md

```diff
@@ -0,0 +1,275 @@
+## Motivation
+
+Storybook & Chromatic provide no way to programmatically define stories for multiple variants. For example, to test the `Button` component with every combination of variant, size, darkMode, icon position etc., we would have to explicitly write each of these stories in the `*.stories.tsx file`, resulting in a lot of manual overhead work. In addition, testing each of these individually would result in hundreds, or even thousands, of stories which would dramatically increase the cost of Chromatic, and generates noise on the sidebar.
+
+As an alternative, this custom decorator reads metadata from the storybook default export, and generates a single story that renders all combinations of props defined in the story metadata.
+
+# Getting started
+
+## Storybook Config
+
+Add a `generate` parameter in the story meta object.
+
+```tsx
+// Button.stories.tsx
+const meta: StoryMeta<typeof Button> = {
+  component: Button,
+  parameters: {
+    default: 'LiveExample',
+    // Define combinations of props to generate
+    generate: {
+      props: {
+        darkMode: [false, true],
+        rightGlyph: [undefined, <Icon glyph={'ArrowRight'} />],
+        leftGlyph: [undefined, <Icon glyph={'Cloud'} />],
+        children: ['MongoDB', undefined],
+        size: Object.values(Size),
+        variant: Object.values(Variant),
+      },
+    },
+  },
+};
+export default meta;
+```
+
+Export an empty story titled `Generated` from the story file
+
+```tsx
+// Button.stories.tsx
+...
+export const Generated = () => {};
+```
+
+This will create a story called "Generated", that renders all defined prop combinations.
+
+![Decorator Demo](./docs/decorator-demo.png)
+
+## Multiple generated stories
+
+You can also break up the generated prop combinations into multiple stories.
+
+To do this, list the generated story names in the `generate` parameter, and set the args you want to remain static on the individual story.
+
+```tsx
+const meta: StoryMeta<typeof Button> = {
+  ...
+  parameters: {
+    generate: {
+      stories: ['LargeSize', 'DefaultSize'],
+      props: { ... },
+    },
+  }
+}
+
+...
+
+export const DefaultSize: StoryType<typeof Button> = () => <></>;
+DefaultSize.parameters = {
+  generate: {
+    args: {
+      size: Size.Default,
+    },
+  },
+};
+
+export const LargeSize: StoryType<typeof Button> = () => <></>;
+LargeSize.parameters = {
+  generate: {
+    args: {
+      size: Size.Large,
+    },
+  },
+};
+```
+
+# Options
+
+## `args`
+
+`Record<key in keyof ComponentProps, ComponentProps[key]>`
+
+By default the decorator pulls args from the main `meta.args` value. However, you may need to set args that are different for the generated instances. In this case,
+use `generate.args` to pass static props that are different from the main story args.
+
+```tsx
+const meta: StoryMeta<typeof Tooltip> = {
+  component: Tooltip,
+  parameters: {
+    default: 'LiveExample',
+    generate: {
+      props: { ... },
+      args: {open: true}
```

Comment:
I'm not entirely clear on the difference between props and args

Reply:
Maybe change `props` to `combine` or `combinations`?
I think `args` needs to stay for consistency with the rest of the storybook api.

@shaneeza @spark33  thoughts?

Reply:
I think it's cleaner to use props for everything and if we're not offering multiple values for a given prop that's okay.

Reply:
I'd expect to be able to specify what properties should remain static as a part of the props definition. Why would it be preferred to specify this through args, rather than simply setting a single value within the prop values arrays? See my example with `darkMode` below:
```js
const meta: StoryMeta<typeof Button> = {
  component: Button,
  parameters: {
    default: 'LiveExample',
    // Define combinations of props to generate
    generate: {
      props: {
        darkMode: [true], // This prop value is unchanging through all instances
        rightGlyph: [undefined, <Icon glyph={'ArrowRight'} />],
        leftGlyph: [undefined, <Icon glyph={'Cloud'} />],
        children: ['MongoDB', undefined],
        size: Object.values(Size),
        variant: Object.values(Variant),
      },
    },
  },
};
export default meta;
```

Reply:
This feels a little odd to me:
```
      props: {
        open: [true],
        darkMode: [false, true],
        baseFontSize: Object.values(BaseFontSize),
        align: Object.values(Align),
        justify: Object.values(Justify),
      },
```

Reply:
I think I'm with Adam on this one. `props` and `args` are sometimes used interchangably in vernacular, but by definition, `properties` are what is accepted, and `arguments` are what's actually passed. It's also how Storybook uses the words. I think the DX is cleaner when you separate the two, and the words aren't confusing to me personally.